### PR TITLE
Assertion rewrite breaks for objects that reimplement `__getattr__`

### PR DESCRIPTION
**Pull request by AnjoMan**
This fix addresses #4631.

### CODE REVIEW

File: testing/test_assertrewrite.py [hunk 1]
```diff
@@ -180,6 +180,27 @@ class X(object):
 
         assert getmsg(f, {"cls": X}) == "assert cls == 42"
 
+    def test_dont_rewrite_if_hasattr_fails(self):
```
Comment (nicoddemus) [inline testing/test_assertrewrite.py]: Interesting, this test passes on `features` just fine on `py36`, but indeed fails on `py27`. 👍

File: 4631.bugfix.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Don't rewrite assertion when __getattr__ is broken
```
Comment (nicoddemus) [inline 4631.bugfix.rst]: Small nitpick:

```rst
Don't rewrite assertion when ``__getattr__`` is broken.
```

File: testing/test_assertrewrite.py (hunk 1 above)
Comment (asottile) [reply testing/test_assertrewrite.py]: This is due to the hasattr changes in python 3 - - perhaps we should make a note here so this gets cleaned up when dropping python 2

File: 4631.bugfix.rst (hunk 2 above)
Comment (AnjoMan) [reply 4631.bugfix.rst]: Thanks! Just realized i also did not put this file in the `changelog` folder.

File: testing/test_assertrewrite.py (hunk 1 above)
Comment (AnjoMan) [reply testing/test_assertrewrite.py]: Can you explain a bit more about what we want to make a note of? As I understand it, this fixes behaviour that we'd only see after [python 3.2](https://docs.python.org/dev/whatsnew/3.2.html), when `hasattr` stopped catching errors that aren't `AttributeError`.

Comment (asottile) [reply testing/test_assertrewrite.py]: oh right, I was thinking the code was in the other direction -- carry on this is fine 👍

**Comment by AnjoMan** [comment]
@nicoddemus I think this is ready to go now.

**Comment by nicoddemus** [comment]
Yes, thanks again!

**Pull request merged**